### PR TITLE
Check that layout variables aren't unconstrained when writing `cmi`s

### PR DESCRIPTION
**Pull request by ncik-roberts**
**Reviewer: @ccasin**

Raise an error if the compiler tries to write out a cmi that mentions an unconstrained layout variable.

The error looks like this:

```
File "external/owee/owee_marker.mli", line 36, characters 3-65:
36 |    | Traverse : ((Obj.t -> 'acc -> 'acc) -> 'acc -> 'acc) service
        ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
Error: Unconstrained layout variable detected when saving artifacts of compilation to disk.
       Please report this error to the Jane Street compilers team. 
```

This immediately found one bug where layout variables of some existential type variables in GADT constructors of extensible type declarations weren't being defaulted. I fix that bug in this PR and add a regression test.

### Implementation

I check whether the layout variable is `None` or not. For the existing callsite, this is a constant-time check because the sort was just returned from `get`, which can only return `Const _` or `Var None`.

Other things to note:
  * There are two callsites to `for_saving`. I distinguish them with a two-value variant. One of them does the layout variable check, the other doesn't.
  * In order to show a location with the error, I plumb a location through `Subst.typexp`. This ends up being rather noisy in the diff. I could just set global mutable state instead if the reviewer thinks we should be worried about rebase conflicts (but I also don't like global mutable state).
  
### Doubts

I stopped commoning-up layout values and truncating histories at the call to `Subst.with_additional_action` that *isn't* just before saving the cmi file. My guess is that this is ok, but it's possible to keep doing the commoning-up and truncation without also doing the layout check, if the reviewer thinks this is a good idea.

### CODE REVIEW

**Review by ccasin — COMMENTED** [review]
The approach here looks good.  I think a few comments need tweaks, and I'd like to glance at the perf next week before merging (but I think it's very unlikely to be an issue).

File: ocaml/typing/subst.mli
Comment (ncik-roberts) [reply ocaml/typing/subst.mli]: I've attempted to reword based on this suggestion, please go ahead and resolve if you think it looks better.

**Comment by ncik-roberts** [comment]
My most recent commits largely respond to review comments. I also decided to unwind my earlier decision to thread `loc` through `typexp` and instead use a global ref &mdash; this helps avoid modifying the hot path of `typexp` more than absolutely necessary (in the interest of avoiding merge conflicts and avoiding changing the allocation behavior of the hot path, etc.). The new location is only used in a (ostensibly) rare error case that's indicative of a compiler bug, so the stakes for getting it wrong are low.

**Review by ccasin — APPROVED** [review]
This looks good to me.

I don't think you need to wait for the new perf numbers to merge (the old ones are fine and you've only made improvements).

**Comment by lpw25** [comment]
I was curious so had a quick look at this. I don't quite understand why one of the `Prepare_for_saving` constructors is taking a function that seems to be always the same -- and even seems to have a comment that says it should always be the same.

**Comment by ccasin** [comment]
Good point.  The previous version does the same thing, and I thought that generality was part of @goldfirere's plan to eventually record a little more info in the cmi for errors.  But looking close I'm not sure it's required even for that plan.  Richard, what do you think?

**Comment by goldfirere** [comment]
This is where the function gets created:

```
let for_saving s =
  (* CR layouts: it would be better to put all this stuff outside this
     function, but it's in here because we really want to tailor the reason
     to describe the module a symbol is imported from. But RAE's initial
     attempt to do this based on filename caused spurious "inconsistent
     assumption" errors that couldn't immediately be solved. Revisit
     with a better approach. *)
  let reason = Layout.Imported in
  let any = Layout.of_const Any ~why:reason in
  let void = Layout.of_const Void ~why:reason in
  let value = Layout.of_const Value ~why:reason in
  let immediate = Layout.of_const Immediate ~why:reason in
  let immediate64 = Layout.of_const Immediate64 ~why:reason in
  let share_layout lay =
    match Layout.get lay with
    | Const Any -> any
    | Const Void -> void
    | Const Value -> value
    | Const Immediate -> immediate
    | Const Immediate64 -> immediate64
    | Var _ -> lay
  in
  { s with for_saving = Some share_layout; last_compose = None }
```

My original attempt put the filename as a field in `Imported`. Then the function approach is much better motivated. But as the CR says, this didn't pan out on a first attempt (I did not try hard to debug, as I wanted to move to other tasks). I still think somehow storing a provenance for imported entities would be nice, which is why I didn't remove the function infrastructure.

It wouldn't be hard to restore if we think it's just getting in the way for now.

**Pull request merged**
